src/Crypto/Encoding/BIP39.hs

```diff
@@ -158,20 +166,18 @@ type ValidEntropySize (n :: Nat) =
 toEntropy :: forall n csz ba
            . (ValidEntropySize n, ValidChecksumSize n csz, ByteArrayAccess ba)
           => ba
-          -> Maybe (Entropy n)
+          -> Either BIP39Err (Entropy n)
 toEntropy bs
-    | BA.length bs*8 == natValInt (Proxy @n) = Just $ Entropy (BA.convert bs) (checksum @csz bs)
-    | otherwise                              = Nothing
+    | BA.length bs*8 == natValInt (Proxy @n) = Right $ Entropy (BA.convert bs) (checksum @csz bs)
+    | otherwise                              = Left $ BIP39ErrInvalidEntropyLength (natValInt (Proxy @n)) (BA.length bs*8)
 
 toEntropyCheck :: forall n csz ba
                 . (ValidEntropySize n, ValidChecksumSize n csz, ByteArrayAccess ba)
                => ba
                -> Checksum csz
-               -> Maybe (Entropy n)
-toEntropyCheck bs s = case toEntropy bs of
-    Nothing -> Nothing
-    Just e@(Entropy _ cs) | cs == s   -> Just e
-                          | otherwise -> Nothing
+               -> Either BIP39Err (Entropy n)
+toEntropyCheck bs s = toEntropy bs >>= \en@(Entropy _ cs) ->
+  if cs == s then Right en else Left $ BIP39ErrInvalidEntropyChecksum (unChecksum cs) (unChecksum s)
```

Comment:
Why not having both `Checksum` instead of `Word8`? Possibly exporting `unChecksum` for convenience.

Reply:
Checksum has a type parameter — the size in bits. Not sure if one can wrap it in an existential or something… Seemed too complicated on first glance